### PR TITLE
Fix unexpected recommendation

### PR DESCRIPTION
**Pull request by jrapin**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

#947 
What happens:
- 1+1 converges to very small sigma/steps very quickly
- when reaching machine precision, it considers it is reestimating the same point
- points which have several estimations have lower pessimistic loss, even though they are not as good
- the best pessimistic point is returned, which is not good in this case.

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [ ] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by jrapin** [comment]
@teytaud not sure how to solve this, maybe use the average as default to return the value ? this seems to be the expected behavior. Returning the pessimistic best as a default even for non-noisy optimization is confusing

**Comment by teytaud** [comment]
> @teytaud not sure how to solve this, maybe use the average as default to return the value ? this seems to be the expected behavior. Returning the pessimistic best as a default even for non-noisy optimization is confusing

In the deterministic setting yes we should return the average.

File: nevergrad/optimization/base.py [hunk 1]
```diff
@@ -488,7 +488,8 @@ def recommend(self) -> p.Parameter:
         """
         recom_data = self._internal_provide_recommendation()  # pylint: disable=assignment-from-none
         if recom_data is None:
-            return self.current_bests["pessimistic"].parameter
+            name = "minimum" if self.parametrization.descriptors.deterministic_function else "pessimistic"
+            return self.current_bests[name].parameter
```
Comment (jrapin) [inline nevergrad/optimization/base.py:492]: @teytaud this is a drastic change in my opinion, you may want to crosscheck it in the xps

File: nevergrad/optimization/base.py [hunk 2]
```diff
@@ -395,7 +395,7 @@ def _update_archive_and_bests(self, candidate: p.Parameter, loss: tp.FloatLoss)
         # update current best records
         # this may have to be improved if we want to keep more kinds of best losss
 
-        for name in ["optimistic", "pessimistic", "average"]:
+        for name in self.current_bests:
```
Comment (teytaud) [inline nevergrad/optimization/base.py:398]: ah yes cool as a shortcut :-)

File: nevergrad/optimization/base.py (hunk 1 above)
Comment (teytaud) [reply nevergrad/optimization/base.py:492]: I think it is safe but yes I'll check xps... complicated though are the absolute indicators are just new.

File: nevergrad/optimization/test_base.py [hunk 3]
```diff
@@ -152,3 +155,25 @@ def test_naming() -> None:
...
+
+    def __init__(self) -> None:
+        self.min_loss = float("inf")
+
+    def __call__(self, score: int) -> float:
+        self.min_loss = min(score, self.min_loss)
+        return score
+
+
+def test_recommendation_correct() -> None:
```
Comment (teytaud) [inline nevergrad/optimization/test_base.py:171]: A possible test would be the example in the issue.
That said, I find the fix quite cool and convincing.

File: nevergrad/optimization/base.py (hunk 1 above)
Comment (jrapin) [reply nevergrad/optimization/base.py:492]: I guess we can wait a bit to merge if it is safer

Comment (jrapin) [reply nevergrad/optimization/base.py:492]: or I provide a fix for the noisy variants

File: nevergrad/optimization/test_base.py (hunk 3 above)
Comment (jrapin) [reply nevergrad/optimization/test_base.py:171]: Actually that is the same test, just simplified a bit, but it does exactly the same

**Review by teytaud — APPROVED** [review]
